lib/ansible/plugins/lookup/password.py

```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+        new_char = random_generator.choice(chars)
+        if new_char in chars:
+            password.append(new_char)
+
+    return u''.join(password)
+
+def _random_salt():
+    salt_chars = to_text(ascii_letters + digits + './', errors='surrogate_or_strict')
+    return _random_password(length=8, chars=salt_chars)
+
+def _gen_candidate_chars(characters):
+    chars = []
+    for chars_spec in characters:
+        chars.append(to_text(getattr(string, to_native(chars_spec), chars_spec),
+                            errors='surrogate_or_strict'))
+    chars = u''.join(to_text(c) for c in chars).replace(u'"', u'').replace(u"'", u'')
+    return chars
+
+
+def _gen_password(length, chars):
+    chars = _gen_candidate_chars(chars)
+    password = u''.join(random.choice(chars) for _ in range(length))
+    return password
+
+def _create_password_file(b_path):
+    b_pathdir = os.path.dirname(b_path)
+
+    try:
+        makedirs_safe(b_pathdir, mode=0o700)
+    except OSError as e:
+        raise AnsibleError("cannot create the path for the password lookup: %s (error was %s)" % (b_pathdir, str(e)))
+
+def _read_password_file(b_path):
+    content = open(b_path, 'rb').read().rstrip()
+    return content
+
+def _parse_password(b_content):
+    content = to_text(b_content, errors='surrogate_or_strict')
+    password = content
+    salt = None
+
+    try:
+        sep = content.rindex(' salt=')
+    except ValueError:
+        # No salt
+        pass
+    else:
+        salt = password[sep + len(' salt='):]
+        password = content[:sep]
 
-class LookupModule(LookupBase):
+    return content, password, salt
 
-    def random_password(self, length=DEFAULT_LENGTH, chars=C.DEFAULT_PASSWORD_CHARS):
-        '''
-        Return a random password string of length containing only chars.
-        NOTE: this was moved from the old ansible utils code, as nothing
-              else appeared to use it.
-        '''
+def _write_password_file(b_path, content):
+    with open(b_path, 'wb') as f:
+        os.chmod(b_path, 0o600)
+        f.write(content)
 
-        password = []
-        while len(password) < length:
-            new_char = os.urandom(1)
-            if new_char in chars:
-                password.append(new_char)
+def _format_content(password, salt, encrypt):
+    if not encrypt:
+        return password
 
-        return ''.join(password)
+    return u'%s salt=%s' % (password, salt)
 
-    def random_salt(self):
-        salt_chars = ascii_letters + digits + './'
-        return self.random_password(length=8, chars=salt_chars)
+def _read_or_create_password_file(b_path):
+    salt = None
+    plaintext_password = None
+    content = None
 
-    def run(self, terms, variables, **kwargs):
+    if os.path.exists(b_path):
+        b_file_content = _read_password_file(b_path)
+        content, plaintext_password, salt = _parse_password(b_file_content)
+    else:
+        _create_password_file(b_path)
+
+    return content, plaintext_password, salt
```

Comment:
I can add a wrapper function that will call both, but I'd prefer to keep the impls as separate methods.